### PR TITLE
feat: define interfaces for the bulker contract

### LINKED ISSUE #27: Define interfaces for the bulker contract
The [Bulker contract](https://github.com/morpho-org/morpho-aave-v3/blob/dev/src/extensions/BulkerGateway.sol) for Morpho Aave V3 will give the possibility to the end user to batch transactions into a single one.

Todo:
- [ ] Define the specs/interfaces to add to the sdk in order to make calls to the Bulker contract.
- [ ] Define a new TxHandler

### PR DESCRIPTION
**Pull request by mlgarchery**
### Description of change

Closes #27 

### Pull-Request Checklist

- [ ] Code is up-to-date with the `main` branch
- [ ] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ ] The new commits follow conventions outlined in the [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/)

### CODE REVIEW

File: src/txHandler/TransactionHandler.interface.ts [hunk 1]
```diff
@@ -12,24 +13,13 @@ import {
...
-export enum TxHandlerOperation {
-  supply = "supply",
-  borrow = "borrow",
-  withdraw = "withdraw",
-  repay = "repay",
-  swap = "swap",
-  claimRewards = "claimRewards",
-  claimMorphoRewards = "claimMorphoRewards",
-  wrapEth = "wrapEth",
-}
```
Comment (mlgarchery) [inline src/txHandler/TransactionHandler.interface.ts:24]: This was unused ( also unused in the monorepo )

File: src/txHandler/Bulker.TxHandler.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  public handleBulkerTransaction(
+    operations: Bulker.Transactions[], // list d'operation possible
+    options?: TransactionOptions | undefined
+  ): Promise<any> {
+    // TODO: to implement
+    return Promise.resolve();
+  }
+
+  public async handleApproval(token: Token, amount: BigNumber, options?: ApprovalHandlerOptions) {
+    // TODO: to implement (same as the Web3TxHandler ?)
```
Comment (julien-devatom) [inline src/txHandler/Bulker.TxHandler.ts:18]:
```suggestion
    // TODO: to implement (same as the Web3TxHandler ?)
    return Promise.reject("TODO: to implement");
```

File: src/txHandler/Bulker.TxHandler.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  public async handlePermit2Approval(
+    token: Token,
+    amount: BigNumber,
+    deadline: BigNumber,
+    nonce: BigNumber,
+    options?: ApprovalHandlerOptions
+  ) {
+    // TODO: to implement (same as the Web3TxHandler ?)
```
Comment (julien-devatom) [inline src/txHandler/Bulker.TxHandler.ts:28]:
```suggestion
    // TODO: to implement (same as the Web3TxHandler ?)
    return Promise.reject("TODO: to implement");
```

File: src/types/tx.ts [hunk 4]
```diff
@@ -1,5 +1,13 @@
 import { BigNumber, CallOverrides } from "ethers";
 
+import { Address } from ".";
+
+export interface Signature {
+  v: number;
+  r: string;
+  s: string;
+}
```
Comment (julien-devatom) [inline src/types/tx.ts:9]: ethers.js hlready has a signature type: https://docs.ethers.org/v4/api-utils.html#signatures

File: src/txHandler/Base.TxHandler.ts [hunk 5]
```diff
@@ -51,16 +72,11 @@ export default abstract class BaseTxHandler implements ITransactionHandler {
...
-    options?: ApprovalHandlerOptions | undefined
-  ): Promise<any>;
-  abstract handlePermit2Approval(
-    token: Token,
-    amount: BigNumber,
-    deadline: BigNumber,
-    nonce: BigNumber,
+}
+
+export abstract class BaseBulkerTxHandler extends TxHandler implements IBulkerTransactionHandler {
```
Comment (julien-devatom) [inline src/txHandler/Base.TxHandler.ts:77]: Bulker don't have to implement `handleApproval`, just `handleBatchTransactions`

File: src/txHandler/TransactionHandler.interface.ts [hunk 6]
```diff
@@ -45,8 +35,12 @@ export interface ITransactionHandler extends ApprovalHandlerInterface {
     options?: TransactionOptions
   ) => Promise<any>;
 
-  handleWrapEth: (
-    amount: BigNumber,
+  handleWrapEth: (amount: BigNumber, options?: TransactionOptions) => Promise<void>;
+}
+
+export interface IBulkerTransactionHandler extends INotifierManager, ApprovalHandlerInterface {
```
Comment (julien-devatom) [inline src/txHandler/TransactionHandler.interface.ts:41]: The interface is not related to the bulker but to the fact to handle multiple operations. The bulker is an implementation for an EOA. I suggest to rename the interface `IBatchTxHandler` and the function `handleBatchTransactions`. Moreover, it doest have to implement ApprovalHandlerInterface

File: src/txHandler/Base.TxHandler.ts (hunk 5 above)
Comment (julien-devatom) [reply src/txHandler/Base.TxHandler.ts:77]: And having `BaseBulkerTxHandler` + `BulkerTxHandler` is over engineering imo

File: src/types/tx.ts [hunk 7]
```diff
@@ -28,8 +36,141 @@ export enum TransactionType {
...
   repay = "Repay",
 }
-
 export interface ClaimTransaction {
   amount: BigNumber;
   proof: string[];
 }
+
+// Bulker transactions
+export namespace Bulker {
```
Comment (julien-devatom) [inline src/types/tx.ts:45]: Bulker is an implementation

File: src/txHandler/Base.TxHandler.ts (hunk 5 above)
Comment (mlgarchery) [reply src/txHandler/Base.TxHandler.ts:77]: But the BulkerTxHandler still needs to handle token approvals before executing the batch in handleBatchTransactions() ?

File: src/txHandler/Mock.TxHandler.ts [hunk 8]
```diff
@@ -13,9 +13,10 @@ import { delay } from "../utils";
 import { getPermit2Message } from "../utils/permit2";
 
 import { ApprovalHandlerOptions } from "./ApprovalHandler.interface";
-import BaseTxHandler from "./Base.TxHandler";
+import { BaseTxHandler } from "./TxHandler";
+import { IOneTxHandler } from "./TxHandler.interface";
 
-export default class MockTxHandler extends BaseTxHandler {
+export default class MockTxHandler extends BaseTxHandler implements IOneTxHandler {
```
Comment (julien-devatom) [inline src/txHandler/Mock.TxHandler.ts:19]:
```suggestion
export default class MockTxHandler extends BaseTxHandler implements ISimpleTxHandler {
```

File: src/txHandler/TxHandler.interface.ts [hunk 9]
```diff
@@ -45,8 +36,9 @@ export interface ITransactionHandler extends ApprovalHandlerInterface {
...
 
-  handleWrapEth: (
-    amount: BigNumber,
-    options?: TransactionOptions
-  ) => Promise<void>;
+  handleWrapEth: (amount: BigNumber, options?: TransactionOptions) => Promise<void>;
+}
+
+export interface IBatchTxHandler extends IBaseTxHandler {
+  handleBatchTransaction: (operations: any[], options?: TransactionOptions) => Promise<any>;
```
Comment (julien-devatom) [inline src/txHandler/TxHandler.interface.ts:43]:
```suggestion
  handleBatchTransactions: (operations: Operation[], options?: TransactionOptions) => Promise<any>;
```

File: src/txHandler/Bulker.TxHandler.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    return Promise.reject("TODO: to implement");
+  }
```
Comment (julien-devatom) [inline src/txHandler/Bulker.TxHandler.ts:31]: To RM

File: src/txHandler/Base.TxHandler.ts (hunk 5 above)
Comment (mlgarchery) [reply src/txHandler/Base.TxHandler.ts]: Ok removing handleApproval as the bulker only accepts permit2 approval

**Review by julien-devatom — CHANGES_REQUESTED** [review]

**Review by julien-devatom — APPROVED** [review]

**Pull request merged**